File: src/java/grpc/armeria/src/main/java/benchmark/BenchmarkServiceImpl.java
package benchmark;

import dev.benchmark.grpc.BenchmarkServiceGrpc;
import dev.benchmark.grpc.BenchmarkProto;
import io.grpc.Status;
import io.grpc.StatusException;
import io.grpc.stub.StreamObserver;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

public class BenchmarkServiceImpl extends BenchmarkServiceGrpc.BenchmarkServiceImplBase {

    private final DatabaseService dbService;
    private final CacheService cacheService;
    private final String version;

    public BenchmarkServiceImpl(DatabaseService dbService, CacheService cacheService) {
        this.dbService = dbService;
        this.cacheService = cacheService;
        this.version = System.getenv().getOrDefault("APP_VERSION", "1.0.0");
    }

    // Scenario 1: Health check
    @Override
    public void health(BenchmarkProto.HealthRequest request,
                       StreamObserver<BenchmarkProto.HealthResponse> responseObserver) {
        String dbStatus = dbService.healthCheck();
        String cacheStatus = cacheService.healthCheck();

        BenchmarkProto.HealthResponse response = BenchmarkProto.HealthResponse.newBuilder()
                .setStatus("ok")
                .setVersion(version)
                .setTimestamp(Instant.now().toString())
                .setDatabase(dbStatus)
                .setCache(cacheStatus)
                .build();

        responseObserver.onNext(response);
        responseObserver.onCompleted();
    }

    // Scenario 2: JSON serialization (1000 items)
    @Override
    public void getJsonItems(BenchmarkProto.JsonItemsRequest request,
                             StreamObserver<BenchmarkProto.JsonItemsResponse> responseObserver) {
        int limit = request.getLimit() > 0 ? request.getLimit() : 1000;

        BenchmarkProto.JsonItemsResponse.Builder builder = BenchmarkProto.JsonItemsResponse.newBuilder();
        for (int i = 1; i <= limit; i++) {
            builder.addItems(BenchmarkProto.JsonItem.newBuilder()
                    .setId(i)
                    .setUuid(UUID.randomUUID().toString())
                    .setName("Item " + i)
                    .setEmail("user" + i + "@benchmark.com")
                    .setCreatedAt(Instant.now().toString())
                    .setIsActive(i % 2 == 0)
                    .build());
        }

        builder.setCount(limit);
        builder.setTimestamp(Instant.now().toString());

        responseObserver.onNext(builder.build());
        responseObserver.onCompleted();
    }

    // Scenario 3: Simple database query
    @Override
    public void getUser(BenchmarkProto.GetUserRequest request,
                        StreamObserver<BenchmarkProto.UserResponse> responseObserver) {
        try {
            DatabaseService.User user = dbService.getUser(request.getId());
            if (user == null) {
                responseObserver.onError(new StatusException(
                        Status.NOT_FOUND.withDescription("User with id " + request.getId() + " not found")));
                return;
            }

            BenchmarkProto.UserResponse response = BenchmarkProto.UserResponse.newBuilder()
                    .setId(user.getId())
                    .setEmail(user.getEmail())
                    .setFirstName(user.getFirstName())
                    .setLastName(user.getLastName())
                    .setAge(user.getAge())
                    .setCreatedAt(user.getCreatedAt())
                    .build();

            responseObserver.onNext(response);
            responseObserver.onCompleted();
        } catch (Exception e) {
            responseObserver.onError(new StatusException(
                    Status.INTERNAL.withDescription("Database error: " + e.getMessage())));
        }
    }

    // Scenario 4: Complex database query (JOIN + aggregation)
    @Override
    public void getComplexOrders(BenchmarkProto.ComplexOrdersRequest request,
                                 StreamObserver<BenchmarkProto.ComplexOrdersResponse> responseObserver) {
        try {
            int days = request.getDays() > 0 ? request.getDays() : 30;
            List<DatabaseService.UserOrderStat> data = dbService.getComplexOrders(days);

            BenchmarkProto.ComplexOrdersResponse.Builder builder = BenchmarkProto.ComplexOrdersResponse.newBuilder();
            builder.setPeriodDays(days);
            builder.setTotalUsers(data.size());

            for (DatabaseService.UserOrderStat stat : data) {
                builder.addData(BenchmarkProto.UserOrderStats.newBuilder()
                        .setUserId(stat.getUserId())
                        .setUserName(stat.getUserName())
                        .setTotalOrders(stat.getTotalOrders())
                        .setTotalValue(stat.getTotalValue())
                        .setAverageOrderValue(stat.getAverageOrderValue())
                        .build());
            }

            responseObserver.onNext(builder.build());
            responseObserver.onCompleted();
        } catch (Exception e) {
            responseObserver.onError(new StatusException(
                    Status.INTERNAL.withDescription("Database error: " + e.getMessage())));
        }
    }

    // Scenario 5: Cache hit/miss
    @Override
    public void getCacheValue(BenchmarkProto.CacheRequest request,
                              StreamObserver<BenchmarkProto.CacheResponse> responseObserver) {
        String key = request.getKey();
        CacheService.CacheResult result = cacheService.get(key);

        if (result.isHit()) {
            BenchmarkProto.CacheResponse response = BenchmarkProto.CacheResponse.newBuilder()
                    .setKey(key)
                    .setValue(result.getValue())
                    .setCached(true)
                    .setTtl(300)
                    .setTimestamp(Instant.now().toString())
                    .build();

            responseObserver.onNext(response);
            responseObserver.onCompleted();
        } else {
            String value = "value-" + System.currentTimeMillis();
            cacheService.set(key, value, 300);

            BenchmarkProto.CacheResponse response = BenchmarkProto.CacheResponse.newBuilder()
                    .setKey(key)
                    .setValue(value)
                    .setCached(false)
                    .setTtl(300)
                    .setTimestamp(Instant.now().toString())
                    .build();

            responseObserver.onNext(response);
            responseObserver.onCompleted();
        }
    }
}

File: src/kotlin/grpc/spring-grpc/src/main/kotlin/benchmark/BenchmarkServiceImpl.kt
package benchmark

import benchmark.BenchmarkServiceGrpcKt
import io.grpc.Status
import io.grpc.StatusException
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.UUID

@Service
class BenchmarkServiceImpl(
    private val dbService: DatabaseService,
    private val cacheService: CacheService
) : BenchmarkServiceGrpcKt.BenchmarkServiceCoroutineImplBase() {

    private val version = System.getenv("APP_VERSION") ?: "1.0.0"

    // Scenario 1: Health check
    override suspend fun health(request: benchmark.HealthRequest): benchmark.HealthResponse {
        val dbStatus = dbService.healthCheck()
        val cacheStatus = cacheService.healthCheck()

        return benchmark.HealthResponse.newBuilder()
            .setStatus("ok")
            .setVersion(version)
            .setTimestamp(Instant.now().toString())
            .setDatabase(dbStatus)
            .setCache(cacheStatus)
            .build()
    }

    // Scenario 2: JSON serialization (1000 items)
    override suspend fun getJsonItems(request: benchmark.JsonItemsRequest): benchmark.JsonItemsResponse {
        val limit = if (request.limit > 0) request.limit else 1000
        val items = (1..limit).map { i ->
            benchmark.JsonItem.newBuilder()
                .setId(i)
                .setUuid(UUID.randomUUID().toString())
                .setName("Item $i")
                .setEmail("user${i}@benchmark.com")
                .setCreatedAt(Instant.now().toString())
                .setIsActive(i % 2 == 0)
                .build()
        }

        return benchmark.JsonItemsResponse.newBuilder()
            .addAllItems(items)
            .setCount(items.size)
            .setTimestamp(Instant.now().toString())
            .build()
    }

    // Scenario 3: Simple database query
    override suspend fun getUser(request: benchmark.GetUserRequest): benchmark.UserResponse {
        val user = dbService.getUser(request.id)
            ?: throw StatusException(Status.NOT_FOUND.withDescription("User with id ${request.id} not found"))

        return benchmark.UserResponse.newBuilder()
            .setId(user.id)
            .setEmail(user.email)
            .setFirstName(user.firstName)
            .setLastName(user.lastName)
            .setAge(user.age)
            .setCreatedAt(user.createdAt)
            .build()
    }

    // Scenario 4: Complex database query (JOIN + aggregation)
    override suspend fun getComplexOrders(request: benchmark.ComplexOrdersRequest): benchmark.ComplexOrdersResponse {
        val days = if (request.days > 0) request.days else 30
        val data = dbService.getComplexOrders(days)

        val orderStats = data.map { stat ->
            benchmark.UserOrderStats.newBuilder()
                .setUserId(stat.userId)
                .setUserName(stat.userName)
                .setTotalOrders(stat.totalOrders)
                .setTotalValue(stat.totalValue)
                .setAverageOrderValue(stat.averageOrderValue)
                .build()
        }

        return benchmark.ComplexOrdersResponse.newBuilder()
            .setPeriodDays(days)
            .setTotalUsers(orderStats.size)
            .addAllData(orderStats)
            .build()
    }

    // Scenario 5: Cache hit/miss
    override suspend fun getCacheValue(request: benchmark.CacheRequest): benchmark.CacheResponse {
        val key = request.key
        val result = cacheService.get(key)

        return if (result.hit) {
            benchmark.CacheResponse.newBuilder()
                .setKey(key)
                .setValue(result.value!!)
                .setCached(true)
                .setTtl(300)
                .setTimestamp(Instant.now().toString())
                .build()
        } else {
            val value = "value-${System.currentTimeMillis()}"
            cacheService.set(key, value, 300)

            benchmark.CacheResponse.newBuilder()
                .setKey(key)
                .setValue(value)
                .setCached(false)
                .setTtl(300)
                .setTimestamp(Instant.now().toString())
                .build()
        }
    }
}
